File: new_dsa/LinkedList/linkedList.cpp
#include<iostream>
using namespace std;

class Node
{
    public:
    int data;
    Node *nxt = NULL;
};
void push(Node **head,int data){
    Node *temp = *head;
    Node *newNode = new Node();
    newNode->data = data;
    newNode->nxt = NULL;
    if(temp == NULL){
        *head = newNode;
    }
    else{
        while(temp->nxt != NULL){
            temp = temp->nxt;
        }
        temp->nxt = newNode;
    }
}
void insertAfter(Node **head,int data,int after){
    Node *temp = *head;
    while(temp->data != after){
        temp = temp->nxt;
    }
    Node *temp2 = temp->nxt;
    temp->nxt = new Node();
    temp->nxt->data = data;
    temp = temp->nxt;
    temp->nxt = temp2;
}
void insertEnd(Node **head,int data){
    Node *newNode = new Node();
    newNode->data = data;
    if(*head == NULL){
        *head = newNode;
        return;
    }
    Node *temp = *head;
    while(temp->nxt != NULL){
        temp = temp->nxt;
    }
    temp->nxt = newNode;
    return;
}
void Printl(Node *head){
    while(head!=NULL){
        cout << head->data << " ";
        head = head->nxt;
    }
}

Node * Middle(Node *head){
    Node *fast = head;
    Node *slow = head;
    while(fast!=NULL && fast->nxt != NULL){
        slow = slow->nxt;
        fast = fast->nxt;
        // if(fast!=NULL){
            fast = fast->nxt;
        // }
        
    }
    return slow;
}

Node *Reverse(Node *head){
    Node *prev = NULL;
    Node *temp = head;
    while(head!=NULL){
        head = head->nxt;
        temp->nxt = prev;
        prev = temp;
        temp = head;
    }
    return prev;
}

Node *cur = new Node();
// cur->data = -1;
Node *res = cur;
void rev(Node *head){
    if(head->nxt == NULL){
        cur->nxt = head;
        cur = cur->nxt;
        return;
    }
    Node *temp = head;
    rev(head->nxt);
    cur->nxt = temp;
    cur = cur->nxt;
    cur->nxt = NULL;
}

int main(){
    Node *head = NULL;
    push(&head,10);
    push(&head,20);
    push(&head,30);
    insertAfter(&head,50,20);
    insertEnd(&head,440);
    cout << Middle(head)->data << endl;
    Printl(head);
    cout << endl;
    // Node *prev = Reverse(head);
    rev(head);
    Printl(res->nxt);

}

File: DSA/LinkedList/LinkList/LinkedList.java
package LinkList;

class Node {
    int data;
    Node nxt;
    Node(int data){
        this.data = data;
        nxt = null;
    }
}
public class LinkedList{
    public Node head = null;
    public void insertFront(int data){
        Node node = new Node(data);
        node.nxt = head;
        head = node;
    }
    public void insert(int data){
        Node node = new Node(data);
        if(head == null){
            head = node;
        }
        else{
            Node n = head;
            while(n.nxt != null){
                n = n.nxt;
            }
            n.nxt = node;
        }
    }
    public void deleteEnd(){
        Node n = head;
        while(n.nxt.nxt!=null){
            n = n.nxt;
        }
        n.nxt = null;
    }
    public void deleteFront(){
        head = head.nxt;
    }
    public void print(Node p){
        Node n = p;
        System.out.print(n.data);
        n = n.nxt;
        while( n != null ){
            System.out.print("-->");
            System.out.print(n.data);
            n = n.nxt;
        }
    }
    public Node reverse(){
        Node n = head;
        Node prev = null;
        Node cur = n;
        Node next = n.nxt;
        while(cur != null){
            Node temp = cur;
            cur.nxt = prev;
            prev = temp;
            cur = next;
            if(next != null)
                next = next.nxt;
        }
        return prev;
    }
}
// class A{
//     public static void main(String[] args){
//         LinkedList ll = new LinkedList();
//         ll.insert(10);
//         ll.insert(20);
//         ll.insert(30);
//         ll.insert(40);
//         ll.insertFront(30);
//         ll.deleteFront();
//         ll.print(ll.head);
//         System.out.println("\nReversed:");
//         ll.print(ll.reverse());
//     }
// }